Gems/ROS2/Code/Source/Camera/PostProcessing/ROS2ImageEncodingConversionComponent.cpp

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include "ROS2ImageEncodingConversionComponent.h"
+#include <AzCore/Outcome/Outcome.h>
+#include <AzCore/RTTI/RTTIMacros.h>
+#include <AzCore/Serialization/EditContext.h>
+#include <AzCore/Serialization/SerializeContext.h>
+#include <sensor_msgs/msg/detail/image__struct.hpp>
+
+namespace AZStd
+{
+    template<>
+    struct hash<ROS2::EncodingConversion>
+    {
+        size_t operator()(const ROS2::EncodingConversion& data) const
+        {
+            return (static_cast<AZ::u16>(data.encodingIn) << 8) | static_cast<AZ::u16>(data.encodingOut);
+        }
+    };
+} // namespace AZStd
+
+namespace ROS2
+{
+    namespace
+    {
+        const AZStd::unordered_map<ImageEncoding, const char*> ImageEncodingNames = {
+            { ImageEncoding::RGBA8, "rgba8" },
+            { ImageEncoding::RGB8, "rgb8" },
+            { ImageEncoding::Mono8, "mono8" },
+            { ImageEncoding::Mono16, "mono16" },
+        };
+        const AZStd::unordered_map<AZStd::string, ImageEncoding> ImageEncodingFromName = {
+            { "rgba8", ImageEncoding::RGBA8 },
+            { "rgb8", ImageEncoding::RGB8 },
+            { "mono8", ImageEncoding::Mono8 },
+            { "mono16", ImageEncoding::Mono16 },
+        };
+
+        void Rgba8ToRgb8(sensor_msgs::msg::Image& image)
+        {
+            const std::string inputEncoding = ImageEncodingNames.at(ImageEncoding::RGBA8);
+            const std::string outputEncoding = ImageEncodingNames.at(ImageEncoding::RGB8);
+            AZ_Assert(image.encoding == inputEncoding, "Image encoding is %s, expected %s", image.encoding.c_str(), inputEncoding.c_str());
+            AZ_Assert(image.step == image.width * 4, "Image step is not width * 4");
+            AZ_Assert(image.data.size() == image.step * image.height, "Image data size is not step * height");
+
+            // Perform conversion in place
+            for (size_t pixelId = 0; pixelId < image.width * image.height; ++pixelId)
+            {
+                size_t pixelOffsetIn = pixelId * 4;
+                size_t pixelOffsetOut = pixelId * 3;
+                image.data[pixelOffsetOut] = image.data[pixelOffsetIn];
+                image.data[pixelOffsetOut + 1] = image.data[pixelOffsetIn + 1];
+                image.data[pixelOffsetOut + 2] = image.data[pixelOffsetIn + 2];
+            }
+            image.encoding = outputEncoding;
+            image.step = image.width * 3;
+            image.data.resize(image.step * image.height);
```

Comment:
Shouldn't the resize of the image data vector occur before the loop that populates it above?

Reply:
during population loop we read from [0, oldSize) and write to [0, newSize].
According to [`cppreference`](https://en.cppreference.com/w/cpp/container/vector/resize) `Vector capacity is never reduced when resizing to smaller size ...`
resizing before the loop is correct but more confusing (reading from outsize the vector size)